pilot/pkg/networking/plugin/authz/rbac.go

```diff
@@ -46,19 +46,91 @@ const (
 	// RbacConfigName is the name of the RbacConfig custom resource that controls the RBAC behavior.
 	RbacConfigName = "rbac-config"
 
-	sourceIP        = "source.ip"
-	sourceService   = "source.service"
-	destinationIP   = "destination.ip"
-	destinationPort = "destination.port"
-	userHeader      = ":user"
-	methodHeader    = ":method"
-	pathHeader      = ":path"
-	//TODO(yangminzhu): Update the key after the final header name of service is decided.
-	serviceHeader       = ":service"
-	requestHeaderPrefix = "request.header["
-	requestHeaderSuffix = "]"
+	// attributes that could be used in both ServiceRole and ServiceRoleBinding.
+	attributeIP     = "ip"
+	attributeHeader = "header:" // header name is after the colon, e.g. "header:User-Agent".
+
+	// attributes that could be used in a ServiceRole constraint.
+	attributePort      = "port"
+	attributeLabel     = "label:"          // label name is after the colon, e.g. "label:version".
+	attributeName      = "name"            // short service name, e.g. "productpage".
+	attributeNamespace = "namespace"       // e.g. "default".
+	attributeSA        = "service-account" // e.g. "bookinfo-productpage"
+
+	// TODO(yangminzhu): Add authn related attributes.
+
+	userHeader   = ":user"
+	methodHeader = ":method"
+	pathHeader   = ":path"
 )
 
+// serviceMetadata is a collection of different kind of information about a service.
+type serviceMetadata struct {
+	name       string            // full qualified service name.
+	labels     map[string]string // labels of the service instance
+	attributes map[string]string // additional attributes of the service
+}
+
+func createServiceMetadata(attr *model.ServiceAttributes, in *model.ServiceInstance) *serviceMetadata {
+	return &serviceMetadata{
+		name:   in.Service.Hostname.String(),
+		labels: in.Labels,
+		attributes: map[string]string{
+			attributeName:      attr.Name,
+			attributeNamespace: attr.Namespace,
+			attributeSA:        in.ServiceAccount,
+		},
+	}
+}
+
+// shouldMatchAttribute returns true if the given attribute should be matched just in the plugin.
+// This is because we already have enough information to finish this in the plugin.
+func shouldMatchAttribute(attr string) bool {
+	switch attr {
+	case attributeName, attributeNamespace, attributeSA:
+		return true
+	}
+	return strings.HasPrefix(attr, attributeLabel)
+}
+
+// match checks if the service is matched to the given rbac access rule.
+// It returns true if the service is matched to the service name and constraints specified in the
+// access rule.
+func (service serviceMetadata) match(rule *rbacproto.AccessRule) bool {
+	if rule == nil {
+		return true
+	}
+
+	// Check if the service name is matched.
+	if !stringMatch(service.name, rule.Services) {
+		return false
+	}
+
+	// Check if the constraints are matched.
+	for _, constraint := range rule.Constraints {
+		if !shouldMatchAttribute(constraint.Key) {
```

Comment:
constraint.Key should be the full attribute names like "request.auth.principal". How does it match to the constants "name"/"namespace" as used in shouldMatchAttribute()?

Reply:
The constraint.Key is only the ServiceRole which we compare right in the pilot, I think what you mean is for the properties in ServiceRoleBinding, that part we'll use the full name and it will be in a later PR, still waiting for the upstream Envoy PR.